### DISCUSSION TITLE
how to use try except when iterating (query) result?

### DISCUSSION THREAD

**Discussion by yitistica** [sqlalchemy/sqlalchemy#8935] [Usage Questions]
Issue: I have specified a sqlalchemy data model with a choice column that only has a subset of choices that present in the actual database table, therefore it will return an error when it encounters a record that sqlalchemy couldn't covert its choice attribute. For example, 

```python
class ChoicesWithoutC(enum.Enum):
    key_a = 'value_a'
    key_b = 'value_b'

class DataModel(base):
    __tablename__ = 'table_a'
    id = Column('id', Integer, primary_key=True, autoincrement=True)
    choice_type = Column('choice_type', ChoiceType(ChoicesWithoutC, impl=String(64)), nullable=False)

# query:
stmt = select(DataModel.choice_type)
results = session.execute(stmt)
results = [i for i in results]   # !!! raise  ValueError: 'value_c' is not a valid ChoicesWithoutC.
```
However, there is a choice 'value_c' in the already created data table. When I executed a query statement it will raise the error ValueError: 'value_c' is not a valid ChoicesWithoutC. I simply want to skip those fail-to-parse records and log it. How could I achieve that? (I try using while loop with nexr(), but it seems even before the iteration of the result begins, the error is prompted).

---------------------------------------------------------------------------------------------------------
to reproduce the issue:

step 1: set up the "already created" table with error prompting record `value_c` in it:

```python

import enum
import sqlalchemy
from sqlalchemy.orm import declarative_base
from sqlalchemy import Column, Integer, String
from sqlalchemy_utils import ChoiceType
from sqlalchemy.orm import sessionmaker


SQLITE_FILE_PATH = 'sqlite:///./sqlite_db.db'

base = declarative_base()


class ChoicesWithC(enum.Enum):  # to set up the dataset;
    key_a = 'value_a'
    key_b = 'value_b'
    key_c = 'value_c'


data = [{'choice_type': 'value_a'},
        {'choice_type': 'value_c'},
        {'choice_type': 'value_b'}]


def inject_data():
    engine = sqlalchemy.create_engine(SQLITE_FILE_PATH)

    class DataModel(base):
        __tablename__ = 'table_a'
        id = Column('id', Integer, primary_key=True, autoincrement=True)
        choice_type = Column('choice_type', ChoiceType(ChoicesWithC, impl=String(64)), nullable=False)

    base.metadata.create_all(bind=engine)
    Session = sessionmaker(engine)

    with Session() as session:
        session.bulk_insert_mappings(DataModel, data, return_defaults=False)
        session.commit()

if __name__ == '__main__':
    inject_data()
```

The above will create a SQLite db with a table `table_a` in it and multiple records.

---------------------------
step 2: in a different session, try to 

```python

import enum
import sqlalchemy
from sqlalchemy.orm import declarative_base
from sqlalchemy import Column, Integer, String
from sqlalchemy_utils import ChoiceType
from sqlalchemy.orm import sessionmaker
from sqlalchemy import select


SQLITE_FILE_PATH = 'sqlite:///./sqlite_db.db'

base = declarative_base()


class ChoicesWithoutC(enum.Enum):  # different
    key_a = 'value_a'
    key_b = 'value_b'


class DataModel(base):
    __tablename__ = 'table_a'
    id = Column('id', Integer, primary_key=True, autoincrement=True)
    choice_type = Column('choice_type', ChoiceType(ChoicesWithoutC, impl=String(64)), nullable=False)


def iter_try_get_result(results):

    while True:
        try:
            result = next(results)
            print(result, 'successful')
        except StopIteration:
            print('stop iteration')
            break
        except Exception as e:
            print('error', e)
            continue


def query_data():

    engine = sqlalchemy.create_engine(SQLITE_FILE_PATH)
    base.metadata.create_all(bind=engine)
    Session = sessionmaker(engine)

    stmt = select(DataModel.choice_type)  # select only second column;

    with Session() as session:
        results = session.execute(stmt)

        # iter method 1:
        results = results.all()  # ValueError: 'value_c' is not a valid ChoicesWithoutC
        
        # iter method 2:
        results = [i for i in results]  # ValueError: 'value_c' is not a valid ChoicesWithoutC
        
        # iter method 3 (custom):  
        iter_try_get_result(results)  # iteration is stopped as soon as next() is called;

if __name__ == '__main__':
    query_data()   # it will return the error;
```

**Comment by lelit** [comment]
OOC, why can't you inject a proper `where` condition, instead of looping thru all the dataset waiting for errors?

**Reply by yitistica** [reply]
because the where condition has an impact on queries performance, it is meant to be a column in a large time-series table, which isn't indexed, (but i abstracted the issue for easier reproducibility in sqlite). I am meant to read every record in the table as well. And I cannot control what is being put into the table, these records (potentially unkown) are a result of some legacy process. 

I think at this point I just have to loop through the returned result and log abnormal records if found, and considering I need to also perform row-wise transformation at the same time, this seem to be a valid approach, all I need is a way to bypass that error.

**Reply by lelit** [reply]
I see. 

By "_I am meant to read every record in the table_" do you mean that effectively you are going to process in some way the "not-parsable" records?

**Reply by yitistica** [reply]
Yes. I will have to perform some mapping on other columns. Then use this choice column to build pivot tables and so on (therefore I have to limit the choices using ChoiceType so that I don't have some unexpected strings and affect the later processes).

**Reply by lelit** [reply]
Ok, then what about keeping that `choice_type` column as a `str` column (possibly "private", with a leading underscore, and using the `name` option to set the right name on the DB), and exposing it thru a `@property`, where the actual conversion to the `Enum` happens?

**Reply by yitistica** [reply]
I have checked out the method you suggested, and I found [hybrid property](https://docs.sqlalchemy.org/en/20/orm/extensions/hybrid.html) that 'sort of' does the job. 

The implementation:

```python
class DataModel(base):
    __tablename__ = 'table_a'
    id = Column('id', Integer, primary_key=True, autoincrement=True)
    _choice_type = Column('choice_type', String(64), nullable=False)

    @hybrid_property
    def choice_type(self):
        try:
            choice = ChoicesWithoutC(self._choice_type)
            return choice
        except Exception as e:
            return None

    @choice_type.expression
    def choice_type(cls):
        return cls._choice_type
```

However, when running the query, the returned rows object is still of the value of `_choice_type` rather than the value of `choice_type` (it is not converted to the hybrid property value automatically) if I run `select(DataModel.choice_type)`; but it will return the right hybrid property value if I run `select(DataModel)`:

```python
def query_data():

    engine = sqlalchemy.create_engine(SQLITE_FILE_PATH)
    base.metadata.create_all(bind=engine)
    Session = sessionmaker(engine)

    # case 1 select only the column:
    stmt = select(DataModel.choice_type)
    with Session() as session:
        results = session.execute(stmt)

        for result in results:
            print(result)
    """
    returns:
    ('value_a',)
    ('value_c',)
    ('value_b',)
    """

    # case 2 select the object:
    stmt = select(DataModel) 
    with Session() as session:
        results = session.execute(stmt)

        for result in results:
            print(result[0].choice_type)
    """
    returns:
    ChoicesWithoutC.key_a
    None
    ChoicesWithoutC.key_b
    """

```

**Reply by lelit** [reply]
Yes, reasonable: in the former case SA is actually using its _core_ layer, and indeed it returns tuples, and the Python-side `@hybrid_property` getter is never called.

**Comment by zzzeek** [comment]
I would use a case statement

```sql
stmt = select(case(~DataModel.choice_type.in_(['choicea', 'choiceb']), None), else_=DataModel.choice_type)) 
```

https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.case

**Reply by yitistica** [reply]
This would work as well! Yet I believe there is a missing parenthesis in the syntax:

```python
class ChoicesWithoutC(enum.Enum):  # for
    key_a = 'value_a'
    key_b = 'value_b'

class DataModel(base):
    __tablename__ = 'table_a'
    id = Column('id', Integer, primary_key=True, autoincrement=True)
    choice_type = Column('choice_type', ChoiceType(ChoicesWithoutC, impl=String(64)), nullable=False)

def query_data():

    engine = sqlalchemy.create_engine(SQLITE_FILE_PATH)
    base.metadata.create_all(bind=engine)
    Session = sessionmaker(engine)

    stmt = select(case(
            (~DataModel.choice_type.in_([choice.value for choice in ChoicesWithoutC]), None),
            else_=DataModel.choice_type
        ))

    with Session() as session:
        results = session.execute(stmt)

        for result in results:
            print(result)

"""
returns:
('value_a',)
(None,)
('value_b',)
"""
```

I notice that the query on columns defined with `ChoiceType`  from sqlalchemy_utils would still automatically parse the string into an enum choice object, i.e., 

```python
    bundle = Bundle('all', DataModel.choice_type)
    stmt = select(bundle)
```
Is it possible we could also get it automatically parsed with `case` combined with `Bundle`?

**Reply by zzzeek** [reply]
you can put a case() expression in a Bundle, say `Bundle('all', case(... DataModel.choice_type ...).label("choice_type"))`
